File: src/types/magicalActions/AnimistPower.ts
import * as DB from "tsondb/schema/dsl"
import { name_in_library } from "../_Activatable.js"
import { OldParameter } from "../_ActivatableSkill.js"
import { CheckResultBasedDuration, DurationUnitValue } from "../_ActivatableSkillDuration.js"
import { ActivatableSkillEffect } from "../_ActivatableSkillEffect.js"
import { PropertyIdentifier, TribeIdentifier } from "../_Identifier.js"
import { ImprovementCost } from "../_ImprovementCost.js"
import { AnimistPowerPrerequisites } from "../_Prerequisite.js"
import { ResponsiveTextOptional } from "../_ResponsiveText.js"
import { SkillCheck } from "../_SkillCheck.js"
import { NestedTranslationMap } from "../Locale.js"
import { Errata } from "../source/_Erratum.js"
import { PublicationRefs, src } from "../source/_PublicationRef.js"

export const AnimistPower = DB.Entity(import.meta.url, {
  name: "AnimistPower",
  namePlural: "AnimistPowers",
  type: () =>
    DB.Object({
      check: DB.Required({
        comment: "Lists the linked three attributes used to make a skill check.",
        type: DB.IncludeIdentifier(SkillCheck),
      }),
      parameters: DB.Required({
        comment: "Measurable parameters of an animist power.",
        type: DB.IncludeIdentifier(AnimistPowerPerformanceParameters),
      }),
      property: DB.Required({
        comment: "The associated property.",
        type: PropertyIdentifier(),
      }),
      tribe_tradition: DB.Required({
        comment: `The tribe traditions the animist power is available to. It may be available to all or only specific tribes.

If no tribe tradition is given, the animist power is generally available to all tribe traditions.`,
        type: DB.Array(TribeIdentifier(), { uniqueItems: true }),
      }),
      improvement_cost: DB.Required({
        comment: "States which column is used to improve the skill.",
        type: DB.IncludeIdentifier(AnimistPowerImprovementCost),
      }),
      prerequisites: DB.Optional({
        comment: "The prerequisites for the animist power.",
        type: DB.IncludeIdentifier(AnimistPowerPrerequisites),
      }),
      levels: DB.Optional({
        comment:
          "The animist power can have multiple levels. Each level is skilled separately. A previous level must be on at least 10 so that the next higher level can be activated and skilled. A higher level cannot be skilled higher than a lower level. Each level also adds an effect text to the text of the first level.",
        type: DB.Array(DB.IncludeIdentifier(AnimistPowerLevel), { minItems: 1 }),
      }),
      src,
      translations: NestedTranslationMap(
        DB.Required,
        "AnimistPower",
        DB.Object({
          name: DB.Required({
            comment: "The animist power’s name.",
            type: DB.String({ minLength: 1 }),
          }),
          name_in_library,
          effect: DB.Required({
            comment:
              "The effect description may be either a plain text or a text that is divided by a list of effects for each quality level. It may also be a list for each two quality levels.",
            type: DB.IncludeIdentifier(ActivatableSkillEffect),
          }),
          cost: DB.Optional({
            isDeprecated: true,
            type: DB.IncludeIdentifier(OldParameter),
          }),
          duration: DB.Optional({
            isDeprecated: true,
            type: DB.IncludeIdentifier(OldParameter),
          }),
          prerequisites: DB.Optional({
            isDeprecated: true,
            type: DB.String({ minLength: 1 }),
          }),
          errata: DB.Optional({
            type: DB.IncludeIdentifier(Errata),
          }),
        }),
      ),
    }),
  instanceDisplayName: {},
  uniqueConstraints: [
    {
      entityMapKeyPath: "translations",
      keyPathInEntityMap: "name",
    },
  ],
})

const AnimistPowerLevel = DB.TypeAlias(import.meta.url, {
  name: "AnimistPowerLevel",
  type: () =>
    DB.Object({
      level: DB.Required({
        comment: "The level number.",
        type: DB.Integer({ minimum: 2 }),
      }),
      src: DB.Optional({
        comment: "The source references, if different than the references for level 1.",
        type: DB.IncludeIdentifier(PublicationRefs),
      }),
      translations: NestedTranslationMap(
        DB.Required,
        "AnimistPowerLevel",
        DB.Object({
          effect: DB.Required({
            comment: "An additional effect description for this level.",
            type: DB.String({ minLength: 1, isMarkdown: true }),
          }),
        }),
      ),
    }),
})

const AnimistPowerPerformanceParameters = DB.Enum(import.meta.url, {
  name: "AnimistPowerPerformanceParameters",
  comment: "Measurable parameters of an animist power.",
  values: () => ({
    OneTime: DB.EnumCase({ type: DB.IncludeIdentifier(OneTimeAnimistPowerPerformanceParameters) }),
    Sustained: DB.EnumCase({
      type: DB.IncludeIdentifier(SustainedAnimistPowerPerformanceParameters),
    }),
  }),
})

const OneTimeAnimistPowerPerformanceParameters = DB.TypeAlias(import.meta.url, {
  name: "OneTimeAnimistPowerPerformanceParameters",
  type: () =>
    DB.Object({
      cost: DB.Required({
        comment:
          "The AE cost value, either a flat value or defined dynamically by the primary patron.",
        type: DB.IncludeIdentifier(OneTimeAnimistPowerCost),
      }),
      duration: DB.Required({
        comment: "The duration.",
        type: DB.IncludeIdentifier(OneTimeAnimistPowerDuration),
      }),
    }),
})

const OneTimeAnimistPowerCost = DB.Enum(import.meta.url, {
  name: "OneTimeAnimistPowerCost",
  values: () => ({
    Fixed: DB.EnumCase({ type: DB.IncludeIdentifier(FixedAnimistPowerCost) }),
    ByPrimaryPatron: DB.EnumCase({ type: DB.IncludeIdentifier(AnimistPowerCostByPrimaryPatron) }),
  }),
})

const FixedAnimistPowerCost = DB.TypeAlias(import.meta.url, {
  name: "FixedAnimistPowerCost",
  type: () =>
    DB.Object({
      value: DB.Required({
        comment: "The (temporary) AE cost value.",
        type: DB.Integer({ minimum: 1 }),
      }),
      interval: DB.Optional({
        comment: "If defined, half of the AE cost `value` has to be paid each interval.",
        type: DB.IncludeIdentifier(DurationUnitValue),
      }),
    }),
})

const AnimistPowerCostByPrimaryPatron = DB.TypeAlias(import.meta.url, {
  name: "AnimistPowerCostByPrimaryPatron",
  type: () =>
    DB.Object({
      interval: DB.Optional({
        comment: "If defined, half of the AE cost `value` has to be paid each interval.",
        type: DB.IncludeIdentifier(DurationUnitValue),
      }),
      translations: NestedTranslationMap(
        DB.Optional,
        "AnimistPowerCostByPrimaryPatron",
        DB.Object({
          note: DB.Required({
            comment: "A note, appended to the generated string in parenthesis.",
            type: DB.IncludeIdentifier(ResponsiveTextOptional),
          }),
        }),
      ),
    }),
})

const OneTimeAnimistPowerDuration = DB.Enum(import.meta.url, {
  name: "OneTimeAnimistPowerDuration",
  values: () => ({
    Immediate: DB.EnumCase({ type: null }),
    Fixed: DB.EnumCase({ type: DB.IncludeIdentifier(DurationUnitValue) }),
    CheckResultBased: DB.EnumCase({ type: DB.IncludeIdentifier(CheckResultBasedDuration) }),
  }),
})

const SustainedAnimistPowerPerformanceParameters = DB.TypeAlias(import.meta.url, {
  name: "SustainedAnimistPowerPerformanceParameters",
  type: () =>
    DB.Object({
      cost: DB.Required({
        comment:
          "The AE cost value, either a flat value or defined dynamically by the primary patron.",
        type: DB.IncludeIdentifier(SustainedAnimistPowerCost),
      }),
    }),
})

const SustainedAnimistPowerCost = DB.Enum(import.meta.url, {
  name: "SustainedAnimistPowerCost",
  values: () => ({
    Fixed: DB.EnumCase({ type: DB.IncludeIdentifier(FixedAnimistPowerCost) }),
    ByPrimaryPatron: DB.EnumCase({ type: DB.IncludeIdentifier(AnimistPowerCostByPrimaryPatron) }),
  }),
})

const AnimistPowerImprovementCost = DB.Enum(import.meta.url, {
  name: "AnimistPowerImprovementCost",
  values: () => ({
    Fixed: DB.EnumCase({ type: DB.IncludeIdentifier(ImprovementCost) }),
    ByPrimaryPatron: DB.EnumCase({ type: null }),
  }),
})

File: Sources/OptolithDatabaseSchema/Types/magicalActions/AnimistPower.swift
import FileDB

@Model
public struct AnimistPower {

  /// Lists the linked three attributes used to make a skill check.
  @Relationship(SkillCheck)
  let check: SkillCheck.ID

  /// Measurable parameters of an animist power.
  @Relationship(AnimistPowerPerformanceParameters)
  let parameters: AnimistPowerPerformanceParameters.ID

  /// The target category – the kind of creature or object – the skill affects.
  @Relationship(AffectedTargetCategories)
  let target: AffectedTargetCategories.ID

  /// The associated property.
  let property: PropertyIdentifier()
      tribe_tradition: Required({
        comment: `The tribe traditions the animist power is available to. It may be available to all or only specific tribes.

If no tribe tradition is given, the animist power is generally available to all tribe traditions.`,
        type: Array(TribeIdentifier(), { uniqueItems: true }),
      }),

  /// States which column is used to improve the skill.
  @Relationship(AnimistPowerImprovementCost)
  let improvement_cost: AnimistPowerImprovementCost.ID
  /// The prerequisites for the animist power.
  @Relationship(AnimistPowerPrerequisites)
  let prerequisites: AnimistPowerPrerequisites.ID?
  /// The animist power can have multiple levels. Each level is skilled separately. A previous level must be on at least 10 so that the next higher level can be activated and skilled. A higher level cannot be skilled higher than a lower level. Each level also adds an effect text to the text of the first level.
  let levels: Array(IncludeIdentifier(AnimistPowerLevel), { minItems: 1 })?

    /// The publications where you can find the entry.
    let src: PublicationRefs

    /// All translations for the entry, identified by IETF language tag (BCP47).
    @Relationship
    let translations: [String: Translation]

    struct Translation { // AnimistPowerTranslation

        /// The animist power’s name.
        let name: String({ minLength: 1 })
          name_in_library,
          effect: Required({
            comment:
              "The effect description may be either a plain text or a text that is divided by a list of effects for each quality level. It may also be a list for each two quality levels.",
            type: IncludeIdentifier(ActivatableSkillEffect),
          }),
          cost: Optional({
            isDeprecated: true,
            type: IncludeIdentifier(OldParameter),
          }),
          duration: Optional({
            isDeprecated: true,
            type: IncludeIdentifier(OldParameter),
          }),
          prerequisites: Optional({
            isDeprecated: true,
            type: String({ minLength: 1 }),
          }),

        let errata: Errata?
    }
}

@Embedded
public struct AnimistPowerLevel {

  /// The level number.
  let level: Integer({ minimum: 2 })
  /// The source references, if different than the references for level 1.
  @Relationship(PublicationRefs)
  let src: PublicationRefs.ID?

    /// All translations for the entry, identified by IETF language tag (BCP47).
    @Relationship
    let translations: [String: Translation]

    struct Translation { // AnimistPowerLevelTranslation

        /// An additional effect description for this level.
        let effect: String({ minLength: 1, isMarkdown: true })
        })
      ),
  }

/// Measurable parameters of an animist power.
@ModelEnum
public enum AnimistPowerPerformanceParameters {
    case OneTime(IncludeIdentifier(OneTimeAnimistPowerPerformanceParameters))
    case Sustained(IncludeIdentifier(SustainedAnimistPowerPerformanceParameters))
}

@Embedded
public struct OneTimeAnimistPowerPerformanceParameters {

  /// The AE cost value, either a flat value or defined dynamically by the primary patron.
  @Relationship(OneTimeAnimistPowerCost)
  let cost: OneTimeAnimistPowerCost.ID

  /// The duration.
  @Relationship(OneTimeAnimistPowerDuration)
  let duration: OneTimeAnimistPowerDuration.ID
  }

@ModelEnum
public enum OneTimeAnimistPowerCost {
    case Fixed(IncludeIdentifier(FixedAnimistPowerCost))
    case ByPrimaryPatron(IncludeIdentifier(AnimistPowerCostByPrimaryPatron))
}

@Embedded
public struct FixedAnimistPowerCost {

  /// The (temporary) AE cost value.
  let value: Integer({ minimum: 1 })
  /// If defined, half of the AE cost `value` has to be paid each interval.
  @Relationship(DurationUnitValue)
  let interval: DurationUnitValue.ID?
  }

@Embedded
public struct AnimistPowerCostByPrimaryPatron {
  /// If defined, half of the AE cost `value` has to be paid each interval.
  @Relationship(DurationUnitValue)
  let interval: DurationUnitValue.ID?
      translations: NestedLocaleMap(
        Optional,
        "AnimistPowerCostByPrimaryPatronTranslation",
        Object({

        /// A note, appended to the generated string in parenthesis.
        let note: IncludeIdentifier(ResponsiveTextOptional)
        })
      ),
  }

@ModelEnum
public enum OneTimeAnimistPowerDuration {
    case Immediate
    case Fixed(IncludeIdentifier(DurationUnitValue))
    case CheckResultBased(IncludeIdentifier(CheckResultBasedDuration))
}

@Embedded
public struct SustainedAnimistPowerPerformanceParameters {

  /// The AE cost value, either a flat value or defined dynamically by the primary patron.
  @Relationship(SustainedAnimistPowerCost)
  let cost: SustainedAnimistPowerCost.ID
  }

@ModelEnum
public enum SustainedAnimistPowerCost {
    case Fixed(IncludeIdentifier(FixedAnimistPowerCost))
    case ByPrimaryPatron(IncludeIdentifier(AnimistPowerCostByPrimaryPatron))
}

@ModelEnum
public enum AnimistPowerImprovementCost {
    case Fixed(IncludeIdentifier(ImprovementCost))
    case ByPrimaryPatron
}
